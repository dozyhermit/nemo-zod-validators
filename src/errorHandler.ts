import { z } from 'zod';
import type { ZodError } from './types';

export const DEFAULT_ERROR_MESSAGE = 'Validation has failed';

// TECHNICAL DEBT: remove .flatten(), use z.flattenError instead
const getErrors = (
  error: ZodError['error'],
  key: keyof z.core.$ZodFlattenedError<string, unknown>
) => {
  if (typeof error?.flatten !== 'function') {
    return undefined;
  }

  if (Object.values(error?.flatten()?.[key] || {}).length <= 0) {
    return undefined;
  }

  return error.flatten()[key];
};

export const errorHandlerWithSchema = ({ error }: ZodError) =>
  Response.json(
    {
      message: getErrors(error, 'fieldErrors') || DEFAULT_ERROR_MESSAGE,
    },
    { status: 422 }
  );

export const errorHandlerWithSchemaFormErrors = ({ error }: ZodError) =>
  Response.json(
    {
      message: getErrors(error, 'formErrors') || DEFAULT_ERROR_MESSAGE,
    },
    { status: 422 }
  );

export const errorHandlerWithTransform = () =>
  Response.json(
    { message: 'Invalid request', success: false },
    {
      status: 400,
    }
  );
