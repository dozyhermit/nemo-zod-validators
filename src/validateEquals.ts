import { MiddlewareFunctionProps } from '@rescale/nemo';
import { NextResponse } from 'next/server';
import { errorHandlerWithTransform } from './errorHandler';
import type { ValidateWithTransform } from './types';

type ValidateEquals<T> = ValidateWithTransform<T>;

/**
 * Validates a local variable against a variable inside a `NextRequest` request using the `===` operator.
 *
 * @param {T} value The value being verified
 * @param {Function} transform How to retrieve a value out of a request
 * @param {Function} errorHandler A custom error function
 *
 * @example
 * validateEquals<String | null>(env.process.MY_ENV_VAR, (request) => request.headers['MY_HEADER']));
 */
export const validateEquals = <T>({
  value,
  transform,
  errorHandler,
}: ValidateEquals<T>) => {
  return async ({ request }: MiddlewareFunctionProps) => {
    if (value !== transform(request)) {
      if (errorHandler) {
        return errorHandler();
      }

      return errorHandlerWithTransform();
    }

    return NextResponse.next();
  };
};
