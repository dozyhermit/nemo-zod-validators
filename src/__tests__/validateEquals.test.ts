import { validateEquals } from '../validateEquals';
import { errorHandlerEquals } from './mocks/errorHandler';

describe('validateEquals', () => {
  test('passes validation when two values equal', async () => {
    const action = validateEquals<String | null>({
      value: 'Hello World',
      transform: (request) => request.headers.get('HELLO_WORLD'),
    });

    const result = await action(
      {
        // @ts-expect-error type inconsistencies due to mocking function props
        headers: { get: () => 'Hello World' },
      },
      {}
    );

    expect(result.status).toBe(200);
  });

  test('fails validation when two values do not equal', async () => {
    const action = validateEquals<String | null>({
      value: 'Hello World',
      transform: (request) => request.headers.get('HELLO_WORLD'),
    });

    const result = await action(
      {
        // @ts-expect-error type inconsistencies due to mocking function props
        headers: { get: () => 'Invalid' },
      },
      {}
    );

    expect(result.status).toBe(400);
  });

  describe('with custom error handler', () => {
    test('passes validation when two values equal', async () => {
      const action = validateEquals<String | null>({
        value: 'Hello World',
        transform: (request) => request.headers.get('HELLO_WORLD'),
        errorHandlerCustom: errorHandlerEquals,
      });

      const result = await action(
        {
          // @ts-expect-error type inconsistencies due to mocking function props
          headers: { get: () => 'Hello World' },
        },
        {}
      );

      expect(result.status).toBe(200);
    });

    test('fails validation when two values do not equal', async () => {
      const action = validateEquals<String | null>({
        value: 'Hello World',
        transform: (request) => request.headers.get('HELLO_WORLD'),
        errorHandlerCustom: errorHandlerEquals,
      });

      const result = await action(
        {
          // @ts-expect-error type inconsistencies due to mocking function props
          headers: { get: () => 'Invalid' },
        },
        {}
      );

      expect(await result.json()).toMatchSnapshot();
    });
  });
});
