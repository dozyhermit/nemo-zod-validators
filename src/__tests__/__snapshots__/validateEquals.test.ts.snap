// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`validateEquals with custom error handler fails validation when two values do not equal 1`] = `
{
  "message": "Invalid equals!",
}
`;
